app/Http/Controllers/Auth/AdmissionController.php

```diff
@@ -165,54 +168,82 @@ public function updateNote(Request $request, Application $application): Redirect
         return redirect()->back();
     }
 
+    /**
+     * Show the finalize page with final names and statistics
+     */
+    public function indexFinalize(): View
+    {
+        $this->authorize('finalize', Application::class);
+        $admitted = Application::query()->with(['user', 'applicationWorkshops'])->admitted()->get()->sortBy('user.name');
+        $not_admitted = Application::query()->whereNotIn('id', $admitted->pluck('id'))->get();
+        return view('auth.admission.finalize', [
+            'semester' => $this->semester(),
+            'admitted_applications' => $admitted,
+            'users_to_delete' => User::query()
+                ->withoutGlobalScope('verified')
+                ->with('application')
+                ->whereIn('id', $not_admitted->pluck('user_id'))
+                //ignore users with any existing role
+                ->whereDoesntHave('roles')
+                ->orderBy('name')
+                ->get()
```

Comment:
Can we factor out this query and the identical one in `finalize` into a helper function?

Also, this looks prone to race conditions; what happens if someone messes with admission statuses between this page is rendered and the finalize button is pressed?

Reply:
Good point. But in real life I expect that the workshop committees will have a deadline to finalise the statuses. Or everything will be handled by the secretary/sys admins. 
Let's just pay attention to this as I don't have an easy solution now.

Reply:
> Let's just pay attention to this as I don't have an easy solution now.

Sounds good to me. But can you still do something with the code duplication for the query?